Makefile

```
@@ -305,15 +295,26 @@ test-unit: clean-cov generate fmt vet ## Run Unit tests.
 
 ##@ Build
 
+WASM_SHIM = $(PROJECT_PATH)/kuadrant-ratelimit-wasm
```

Can we use a more generic name for this, that perhaps continues to work if we, say, add something like ext authz to the functions performed by the component as well?

```suggestion
WASM_SHIM = $(PROJECT_PATH)/kuadrant-wasm-shim
```

---

If we are planning to have just one wasm shim to do everything, it makes sense @guicassolato suggestion... however, if the name is meant to only rate limiting purposes, it's OK.

---

I'd go for the more generic name too... unless we need to discriminate one day, which I doubt, this is the most portable option.

---

I have no idea what's better between one wasm shim per function (RL, auth) or a single one for all. I can see pros and cons for both, although today I'd probably be more inclined to a single one, I think.

---

My initial idea was that if ext authz was being done by wasm, it would be called `kuadrant-authz-wasm`. And would be a different wasm binary. Discrimination comes from the fact that they are essentially speaking different languages. While rate limiting uses RLS, ext auth uses [external authorization](https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/ext_authz) gRPC protocol. So configuration of each wasm module, I pre-asume, would be different. Not even speaking about potentially different release stream. 

But if you prefer to go for a generic name for now, I am happy with it. It can always be changed in the future.

---

I agree that RLS and ext_authz are indeed what make the two wasm-shims (two functions of a single wasm-shim) different one another.

On the other hand, other than that, arguably the two wasm-shims are practically identical:
- both need to decide whether a request matches;
- both need to evaluate well-known attributes;
- both perform a grpc call to a service, basically saying "should I let traffic go through? decide based on this payload.";
- both expect a boolean response, maybe with some metadata that typically become HTTP headers.

Moreover, if the auth layer needs to propagate data to the RL layer, with one wasm-shim, that can be done "over-the-wire", while with two, one wasm-shim needs to inject Envoy Dynamic Metadata so the other one can retrieve it.

---

as a design pattern, I always pick one module does one thing. For multiple reasons. But this is a discussion we do not need to have now.

---

> one module does one thing

Not wrong. Neither we should repeat ourselves, another design principle 😜

Maybe this is one of those cases where following all the "rules" we have in life becomes impractical, like _"The early bird catches the worm"_ but also _"Good things come to those who wait."_